### PR TITLE
Adds rule weighting

### PR DESCRIPTION
**Pull request by MadmanMartian**
Current PoC is simple animal mobs, where they take into account if the target is an object or a mob, and the distance.

Pass it a list of objects and a list of rules, and it'll return the list re-ordered with highest priority first

May drop the weighting of the object rule some more, just so mobs have total priority.

:cl:
 * tweak: Overhauled simple animal logic. This should improve target decision making and fix some behaviours like getting caught trying to move to an object through a crowd of potential targets.

### CODE REVIEW

**Comment by MadmanMartian** [comment]
additional: Turns out there was a better way

File: code/datums/fuzzy_rules.dm [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		return L
+	var/list/comparison = L.Copy()
+	var/evaluating = TRUE
+	//First, we associate a value with each object in the comparison list
+	for(var/datum/D in comparison)
+		var/relevance
+		for(var/datum/fuzzy_ruling/FR in rules)
+			relevance += FR.evaluate(D)*FR.weighting
+		comparison[D] = relevance
+	//Then, bubblesort
```
Comment (Exxion) [inline code/datums/fuzzy_rules.dm]: Come on, man

Comment (Exxion) [reply code/datums/fuzzy_rules.dm]: Just a tip: Check out `__HELPERS/sorts`

**Comment by Kurfursten** [comment]
Please, please finish/merge this. Spider memes have gone on too long.

Also can you write the changelog in a way that's more readable to non-coder players and add fixes #21128 to the OP? Something like

>Overhauled simple animal logic. This should improve target decision making and fix some behaviors like webbing open bolted doors.

File: code/datums/fuzzy_rules.dm (hunk 1 above)
Comment (MadmanMartian) [reply code/datums/fuzzy_rules.dm]: What am I looking for exactly?

Comment (Exxion) [reply code/datums/fuzzy_rules.dm]: Anything better than bubblesort
Which is anything in that file

Comment (MadmanMartian) [reply code/datums/fuzzy_rules.dm]: okay, may look into timSort, problem is my lists are currently formatted in the wrong way to use any of the comparison procs, so should be fun

Comment (Exxion) [reply code/datums/fuzzy_rules.dm]: If I'm not mistaken, just use the `associative` parameter

Comment (MadmanMartian) [reply code/datums/fuzzy_rules.dm]: That does make a lot of sense

**Comment by Kurfursten** [comment]
I guess I misunderstood because this did not fix spiders meming doors

**Comment by MadmanMartian** [comment]
It was to work towards it, as I was going to move spiders webbing stuff up to their canattack and such, which got me thinking 'they are going to be too focused on every single object to actually attack anyone'

**Comment by Kurfursten** [comment]
Spidergunk is pretty important to make a priority, people are really impatient about it. I'm not urging you on because of me - honestly, it's been a sore spot in the thread for a long time.

**Pull request merged**